proposals/007-plugin-api-to-select-tls-credentials-for-server-connection.md

```diff
@@ -0,0 +1,178 @@
+# Plugin API for selecting TLS client credentials for proxy-to-server connection
+
+This proposes a new plugin API for allowing the selection of the TLS client credentials to be used for the connection between the proxy and a Kafka server.
+It makes use of [proposal-004](proposal-004) for the terminology used.
+
+## Current situation
+
+Currently the end user configures the client TLS information to be used for connections between the proxy and a target cluster.
+This means that the client TLS certificate used cannot depend on information known at runtime. 
+This includes, in particular information about the client, such as the TLS client certificate that it provided to the proxy.
+
+## Motivation
+
+The lack of API support means all proxied clients have the same TLS identity from the broker's point of view. 
+This means a broker using the `TLS` value for `security.protocol` cannot distinguish between those clients.
+
+Goals: 
+
+* Enable the selection of TLS client certificates at runtime.
+
+## Proposal
+
+
+### API for selecting target cluster TLS credentials
+
+We will add a new plugin interface, `ServerTlsCredentialSupplierFactory`.
+It will use the usual Kroxylicious plugin mechanism, leveraging `java.util.Service`-based discovery.
+However, this plugin is not the same thing as a `FilterFactory`.
+Rather, an implementation class will be defined on the `TargetCluster` configuration object, and instantiated once for each target cluster.
+The TargetCluster's `tls` object will gain a `tlsCredentialSupplier` property, supporting `type` and `config` properties (similarly to how filters are configured).
```

Comment:
Sidebar: not for this proposal. Wonder where we would put this in the operator API. In operator land KafkaService has the upstream tls details, but we can reuse a KafkaService across virtual kafka clusters. Would we put it on KafkaService, so users need multiple KafkaServices if they want different selection logic per virtual cluster. Or put it on VirtualCluster, with the downside that we spread out the upstream TLS settings across two CRs.